core/modules/replica/Configuration.h

```diff
@@ -429,75 +430,610 @@ class Configuration: public ConfigurationIFace {
+     *   the method will do noting.
+     */
+    void reload();
+
+    /**
+     * Reload non-static parameters of the Configuration from the given JSON object.
+     * @note If the previous state of the object was configured from a source having
+     *   a persistent back-end (such as MySQL) then the association with the backend
+     *   will be lost upon completion of the method.
+     * @param obj The input configuration parameters.
+     * @throw std::runtime_error If the input configuration is not consistent
+     *   with expectations of the application.
+     */
+    void reload(nlohmann::json const& obj);
+
+    /**
+     * Reload non-static parameters of the Configuration from an external source.
+     * @param configUrl The configuration source,
+     * @throw std::invalid_argument If the URL has unsupported scheme or it couldn't
+     *   be parsed.
+     * @throw std::runtime_error If the input configuration is not consistent with
+     *   expectations of the application.
+     */
+    void reload(std::string const& configUrl);
+    
+    /**
+     * Construct the original (minus security-related info) path to
+     * the configuration source.
+     * @param showPassword If a value of the flag is 'false' then hash a password
+     *   in the result.
+     * @return The constructed path.
+     */
+    std::string configUrl(bool showPassword=false) const;
+
     /**
-     * @param impl A pointer to the actual implementation of the Configuration.
+     * The directory method for locating categories and parameters within
+     * the given category known to the current implementation.
+     * @note The method only returns the so called "general" categories
+     *   of primitive parameters that exclude workers, databa families,
+     *   databases, etc.
+     * @return A collection of categories and parameters within the given category.
+     *   The name of a category would be the dictionary key, and a value of
+     *   the dictionary will contains a set of the parameter names within
+     *   the corresponding category.
      */
-    Configuration(ConfigurationIFace::Ptr const& impl) : _impl(impl) {}
+    std::map<std::string, std::set<std::string>> parameters() const;
+
+    /**
+     * Return a value of the parameter found by its category and its name.
+     * @param category The name of the parameter's category.
+     * @param param The name of the parameter within its category.
+     * @return A value of the parameter of the requested type.
+     * @throws std::invalid_argument If the parameter doesn't exist, or it has unexpected type.
+     */
+    template <typename T>
+    T get(std::string const& category, std::string const& param) const {
+        util::Lock const lock(_mtx, _context(__func__));
+        try {
+            return _get(lock, category, param).get<T>();
+        } catch (std::invalid_argument const&) {
+            throw;
+        } catch (std::exception const& ex) {
+            throw ConfigTypeMismatch(
```

Comment:
This may be true in a strict sense. This exception encapsulates errors reported by the `json::get<t>()`. Exceptions in the JSON library are the grey area. They aren't properly documented. So, I chose to catch them and tag with my own custom exception just to tell clients that there was a problem during type conversion into the proposed type  (template parameter `T`). I don't think any further refinement of the API is needed here. I don't see any significant benefits from that (yet).  This is the code when this is done:
```
    template <typename T>
    T get(std::string const& category, std::string const& param) const {
        util::Lock const lock(_mtx, _context(__func__));
        try {
            return _get(lock, category, param).get<T>();
        } catch (std::invalid_argument const&) {
            throw;
        } catch (std::exception const& ex) {
            throw ConfigTypeMismatch(
                _context(__func__) + " failed to convert the parameter for category '"
                + category + "', and param '" + param + "', ex: " + std::string(ex.what()) + "'.");
        }
    }
```